lib/environ/src/compilation.rs

```diff
@@ -1,24 +1,77 @@
 //! A `Compilation` contains the compiled function bodies for a WebAssembly
 //! module.
 
-use cranelift_codegen::binemit;
-use cranelift_codegen::ir;
-use cranelift_codegen::CodegenError;
+use crate::module;
+use cranelift_codegen::{binemit, ir, isa, CodegenError};
 use cranelift_entity::PrimaryMap;
 use cranelift_wasm::{DefinedFuncIndex, FuncIndex, WasmError};
+use std::ops::Range;
 use std::vec::Vec;
 
+type FunctionRanges = PrimaryMap<DefinedFuncIndex, Range<usize>>;
+
 /// The result of compiling a WebAssembly module's functions.
 #[derive(Debug)]
 pub struct Compilation {
+    buffer: Vec<u8>,
     /// Compiled machine code for the function bodies.
-    pub functions: PrimaryMap<DefinedFuncIndex, Vec<u8>>,
+    functions: FunctionRanges,
 }
 
 impl Compilation {
+    /// Creates a compilation artifact from a contiguous function buffer and a set of ranges
+    pub fn new(buffer: Vec<u8>, functions: FunctionRanges) -> Self {
+        Self { buffer, functions }
+    }
+
     /// Allocates the compilation result with the given function bodies.
-    pub fn new(functions: PrimaryMap<DefinedFuncIndex, Vec<u8>>) -> Self {
-        Self { functions }
+    pub fn from_function_map(functions: PrimaryMap<DefinedFuncIndex, Vec<u8>>) -> Self {
+        let mut out_map = PrimaryMap::new();
+        let mut buffer = vec![];
+
+        for (_, f) in functions.into_iter() {
+            let start = buffer.len();
+            buffer.extend(f);
+            let end = buffer.len();
+            out_map.push(start..end);
+        }
```

Comment:
Ok, so this is taking the functions that Cranelift compiled into separate `Vec`s, and concatenating them into one big `Vec`, and remembering the indices. And this is because that's what Lightbeam produces. I'll need to think about this a bit more :-).

Reply:
Fixed, we now convert Lightbeam's format into Cranelift's format instead of the other way around.